src/cargo/util/context/mod.rs

```diff
@@ -1590,10 +1590,14 @@ impl GlobalContext {
             }
         }
 
+        // Need to check the `$CARGO_HOME/config` have been warned already.
+        let home_config_without_extention = home.join("config");
+        let already_warned = stash.contains(&home_config_without_extention);
+
         // Once we're done, also be sure to walk the home directory even if it's not
         // in our history to be sure we pick up that standard location for
         // information.
-        if let Some(path) = self.get_file_path(home, "config", true)? {
+        if let Some(path) = self.get_file_path(home, "config", !already_warned)? {
```

Comment:
Isn't this issue really about us somtimes processing `CARGO_HOME`s config twice?  Seems like we should organize the fix (including comments) around that concept.

Would that be as simple as `if pwd.starts_with(&home)`?

Reply:
Should avoid processing `CARGO_HOME`s config twice, I mean.

Totally not processing the `CARGO_HOME`s if it had been walked already.

Reply:
> Would that be as simple as if pwd.starts_with(&home)?

It depends on the relation betweent `pwd` and `home`, the `starts_with` works only pwd is `home`'s subdirectory.

And the other problem is `$CARGO_HOME` can be change by env, so we must `join` the actual `home` to retrieve the  config we had walked.

Reply:
I had an idea for another iteration and decided to try it out myself rather than trying to describe it and found I misunderstood this!  I thought it was a problem with what file gets put in `seen`.  It turns out its because `seen` isn't covering enough.  Reusing that check by extending the scope it covers fixes this.  I've pushed that to the branch and would be interested in your thoughts on it.

Reply:
> Your project is inside of `$HOME`

This is exacly the root reason.  

I compiled the branch on local and run the fixed `cargo`, that seems the problem was addressed.